### PR TITLE
Add TypedArray#filter

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

File: std/assembly/typedarray.ts [hunk 1]
```diff
@@ -1364,6 +1408,38 @@ function MAP<TArray extends ArrayBufferView, T>(
...
+    if (fn(value, i, array)) {
+      store<T>(
+        outDataStart + (<usize>j << alignof<T>()),
+        value
+      );
+      ++j;
+    }
+  }
+  // shrink output buffer
+  var finalLenght = <usize>j << alignof<T>();
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: typo: length

File: std/assembly/typedarray.ts [hunk 2]
```diff
@@ -1364,6 +1408,38 @@ function MAP<TArray extends ArrayBufferView, T>(
...
+  var len = array.length;
+  var out = instantiate<TArray>(len);
+  var dataStart    = array.dataStart;
+  var outDataStart = out.dataStart;
+  var j = 0;
+  for (let i = 0; i < len; i++) {
+    let value = load<T>(dataStart + (<usize>i << alignof<T>()));
+    if (fn(value, i, array)) {
+      store<T>(
+        outDataStart + (<usize>j << alignof<T>()),
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: moving `j++` here might save a byte or two. maybe, not sure

File: std/assembly/typedarray.ts [hunk 3]
```diff
@@ -1364,6 +1408,38 @@ function MAP<TArray extends ArrayBufferView, T>(
...
 }
 
+// @ts-ignore: decorator
+@inline
+function FILTER<TArray extends ArrayBufferView, T>(
+  array: TArray,
+  fn: (value: T, index: i32, self: TArray) => bool,
+): TArray {
+  var len = array.length;
+  var out = instantiate<TArray>(len);
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: would be cleaner to allocate a buffer, fill it, realloc it, and then build a `TArray` from it directly. minimal improvement, but saves reassigning it after realloc.

Comment (MaxGraey) [reply std/assembly/typedarray.ts]: Yes. Also it skip unnecessary `memory.fill`

File: std/assembly/typedarray.ts [hunk 4]
```diff
@@ -1415,28 +1415,26 @@ function FILTER<TArray extends ArrayBufferView, T>(
   fn: (value: T, index: i32, self: TArray) => bool,
 ): TArray {
   var len = array.length;
-  var out = instantiate<TArray>(len);
-  var dataStart    = array.dataStart;
-  var outDataStart = out.dataStart;
-  var j = 0;
+  var byteLength = len << alignof<T>();
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: nit: this var seems to be unnecessary

Comment (MaxGraey) [reply std/assembly/typedarray.ts]: Done!

**Comment by dcodeIO** [comment]
Thanks!

**Pull request merged**